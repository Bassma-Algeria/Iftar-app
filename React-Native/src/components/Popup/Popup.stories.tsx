import React, {useState} from 'react';

import {storiesOf} from '@storybook/react-native';

import {Popup} from './Popup';
import {Button, View} from 'react-native';
import {Header} from '../Header/Header';

const PopupStories = storiesOf('Popup', module);

const Component = () => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <View>
      <Button title="toggle popup" onPress={() => setIsOpen(!isOpen)} />
      <Popup isOpen={isOpen}>
        <Header style={{margin: 40}} variant="h1">
          Some Content
        </Header>
      </Popup>
    </View>
  );
};

PopupStories.add('Default', () => <Component />);
