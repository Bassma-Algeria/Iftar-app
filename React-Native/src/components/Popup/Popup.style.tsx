import {Dimensions, StyleSheet} from 'react-native';
import {COLORS} from '../../theme/Colors';

const styles = StyleSheet.create({
  container: {
    backgroundColor: COLORS.beige,
    position: 'absolute',
    bottom: 0,
    left: 0,
    width: '100%',
    borderTopRightRadius: 30,
    borderTopLeftRadius: 30,
    elevation: 40,
    shadowColor: COLORS.primary,
    zIndex: 3,
    maxHeight: Dimensions.get('window').height - 30,
  },

  overlay: {
    position: 'absolute',
    width: '100%',
    height: '100%',
    top: 0,
    backgroundColor: 'black',
  },

  topBar: {
    width: 80,
    height: 6,
    backgroundColor: COLORS.brownShade,
    borderRadius: 3,
    marginTop: 5,
    alignSelf: 'center',
  },
});

export {styles};
