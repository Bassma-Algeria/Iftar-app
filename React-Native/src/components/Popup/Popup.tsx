import React, {useEffect} from 'react';
import {View, Animated, Pressable} from 'react-native';
import {useNavigation} from '@react-navigation/native';

import {styles} from './Popup.style';

import {usePopupAnimation} from './_hooks_/usePopupAnimation';

interface Props {
  isOpen: boolean;
  onClose?: Function;
}

const Popup: React.FC<Props> = ({isOpen, children, onClose}) => {
  const navigation = useNavigation();
  const {translateY, closePopup, openPopup} = usePopupAnimation(onClose);

  useEffect(() => {
    if (isOpen) {
      openPopup();
    } else {
      closePopup();
    }
  }, [closePopup, openPopup, isOpen]);

  useEffect(() => {
    const unsubscribe = navigation.addListener('beforeRemove', e => {
      if (!isOpen) {
        return;
      }

      e.preventDefault();
      closePopup();
    });

    return unsubscribe;
  }, [closePopup, navigation, isOpen]);

  return (
    <>
      {isOpen && <Overlay onPress={closePopup} />}

      <Animated.ScrollView style={[styles.container, {transform: [{translateY}]}]}>
        <View style={styles.topBar} />

        {children}
      </Animated.ScrollView>
    </>
  );
};

interface OverlayProp {
  onPress: () => void;
}

const Overlay: React.FC<OverlayProp> = ({onPress}) => {
  return <Pressable style={styles.overlay} onPress={onPress} />;
};

export {Popup};
