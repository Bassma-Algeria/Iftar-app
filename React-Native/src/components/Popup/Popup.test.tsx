import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import {render} from '@testing-library/react-native';
import {NavigationContainer} from '@react-navigation/native';

import {Popup} from './Popup';
import {Text} from 'react-native';

describe('Popup Component', () => {
  it('should render correclty and closed when isOpen is false', () => {
    const tree = renderer
      .create(
        <NavigationContainer>
          <Popup isOpen={false} setIsOpen={() => {}}>
            <Text>Some content</Text>
          </Popup>
        </NavigationContainer>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should open when isOpen is true', () => {
    const tree = renderer
      .create(
        <NavigationContainer>
          <Popup isOpen setIsOpen={() => {}}>
            <Text>Some content</Text>
          </Popup>
        </NavigationContainer>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should call the onAuthClose the popup is auto closed', () => {
    const onClose = jest.fn();

    render(
      <NavigationContainer>
        <Popup isOpen={false} setIsOpen={() => {}} onAutoClose={onClose}>
          <Text>Some content</Text>
        </Popup>
      </NavigationContainer>,
    );

    expect(onClose).toBeCalledTimes(1);
  });

  it('should be able to change the background color', () => {
    const tree = render(
      <NavigationContainer>
        <Popup isOpen setIsOpen={() => {}} backgroundColor="black">
          <Text>Some content</Text>
        </Popup>
      </NavigationContainer>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
