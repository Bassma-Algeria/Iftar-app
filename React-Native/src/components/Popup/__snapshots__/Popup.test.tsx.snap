// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup Component should be able to change the container style 1`] = `
<View
  collapsable={false}
  nativeID="animatedComponent"
  style={
    Object {
      "backgroundColor": "#F4E6CA",
      "borderTopLeftRadius": 30,
      "borderTopRightRadius": 30,
      "bottom": 0,
      "elevation": 40,
      "height": "auto",
      "left": 0,
      "maxHeight": 1304,
      "position": "absolute",
      "shadowColor": "#B87015",
      "transform": Array [
        Object {
          "translateY": 1334,
        },
      ],
      "width": "100%",
      "zIndex": 3,
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "#B66D12",
        "borderRadius": 3,
        "height": 6,
        "marginTop": 5,
        "width": 80,
      }
    }
  />
  <RCTScrollView
    contentContainerStyle={
      Object {
        "margin": 30,
      }
    }
  >
    <View>
      <Text>
        Some content
      </Text>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Popup Component should open when isOpen is true 1`] = `
<View
  collapsable={false}
  nativeID="animatedComponent"
  style={
    Object {
      "backgroundColor": "#F4E6CA",
      "borderTopLeftRadius": 30,
      "borderTopRightRadius": 30,
      "bottom": 0,
      "elevation": 40,
      "height": "auto",
      "left": 0,
      "maxHeight": 1304,
      "position": "absolute",
      "shadowColor": "#B87015",
      "transform": Array [
        Object {
          "translateY": 1334,
        },
      ],
      "width": "100%",
      "zIndex": 3,
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "#B66D12",
        "borderRadius": 3,
        "height": 6,
        "marginTop": 5,
        "width": 80,
      }
    }
  />
  <RCTScrollView>
    <View>
      <Text>
        Some content
      </Text>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Popup Component should render correclty and closed when isOpen is false 1`] = `
<View
  collapsable={false}
  nativeID="animatedComponent"
  style={
    Object {
      "backgroundColor": "#F4E6CA",
      "borderTopLeftRadius": 30,
      "borderTopRightRadius": 30,
      "bottom": 0,
      "elevation": 40,
      "height": "auto",
      "left": 0,
      "maxHeight": 1304,
      "position": "absolute",
      "shadowColor": "#B87015",
      "transform": Array [
        Object {
          "translateY": 1334,
        },
      ],
      "width": "100%",
      "zIndex": 3,
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "#B66D12",
        "borderRadius": 3,
        "height": 6,
        "marginTop": 5,
        "width": 80,
      }
    }
  />
  <RCTScrollView>
    <View>
      <Text>
        Some content
      </Text>
    </View>
  </RCTScrollView>
</View>
`;
